File: js/two-pointers/container-with-most-water.js
/**
 * https://leetcode.com/problems/container-with-most-water
 * 
 * @param {number[]} height
 * @return {number}
 */
const maxArea = (height) => {

    let curr = 0;

    let lPtr = 0;
    let rPtr = height.length - 1;

    while (lPtr < rPtr) {
        const smallestSide = Math.min(height[lPtr], height[rPtr]);
        const volume = smallestSide * (rPtr - lPtr); 

        curr = Math.max(curr, volume);

        if (height[lPtr] < height[rPtr]) {
            lPtr++;
        } else {
            rPtr--;
        }
    };

   return curr;
};


//Same application as three sum apart from we don't need to sort.
//Key to problem is seeing that can caculate the volume of any two numbers by taking
//the smallest of two * the difference between the indexes

File: python/two-pointers/container-with-most-water.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        
        l = 0
        r = len(heights) - 1

        most_water = (r - l) * min(heights[l], heights[r])

        while (l < r):

            if (heights[l] < heights[r]):
                l += 1
            elif (heights[r] <= heights[l]):
                r -= 1

            most_water = max(most_water, min(heights[l], heights[r]) * (r - l))

        return most_water

            

# https://leetcode.com/problems/container-with-most-water
# Typical two pointer, calculate max area by taking lowest side and total of numbers between two indexes, then times that by lowest side.
# Increment / decincrement pointers based on what's lower / higher